File: implementations/java/common/src/main/java/com/theshuai/common/stun/StunMessage.java
package com.theshuai.common.stun;

import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HexFormat;
import java.util.List;
import java.util.Optional;

public final class StunMessage {
    public static final int MAGIC_COOKIE = 0x2112A442;
    public static final int HEADER_BYTES = 20;
    public static final int TRANSACTION_ID_BYTES = 12;

    public static final int BINDING_REQUEST = 0x0001;
    public static final int BINDING_SUCCESS = 0x0101;
    public static final int BINDING_ERROR = 0x0111;
    public static final int ALLOCATE_REQUEST = 0x0003;
    public static final int ALLOCATE_SUCCESS = 0x0103;
    public static final int ALLOCATE_ERROR = 0x0113;
    public static final int REFRESH_REQUEST = 0x0004;
    public static final int REFRESH_SUCCESS = 0x0104;
    public static final int REFRESH_ERROR = 0x0114;
    public static final int CREATE_PERMISSION_REQUEST = 0x0008;
    public static final int CREATE_PERMISSION_SUCCESS = 0x0108;
    public static final int CREATE_PERMISSION_ERROR = 0x0118;
    public static final int SEND_INDICATION = 0x0016;
    public static final int DATA_INDICATION = 0x0017;

    public static final int ATTR_MAPPED_ADDRESS = 0x0001;
    public static final int ATTR_CHANGE_REQUEST = 0x0003;
    public static final int ATTR_USERNAME = 0x0006;
    public static final int ATTR_MESSAGE_INTEGRITY = 0x0008;
    public static final int ATTR_ERROR_CODE = 0x0009;
    public static final int ATTR_UNKNOWN_ATTRIBUTES = 0x000A;
    public static final int ATTR_LIFETIME = 0x000D;
    public static final int ATTR_XOR_PEER_ADDRESS = 0x0012;
    public static final int ATTR_DATA = 0x0013;
    public static final int ATTR_REALM = 0x0014;
    public static final int ATTR_NONCE = 0x0015;
    public static final int ATTR_XOR_RELAYED_ADDRESS = 0x0016;
    public static final int ATTR_REQUESTED_TRANSPORT = 0x0019;
    public static final int ATTR_XOR_MAPPED_ADDRESS = 0x0020;
    public static final int ATTR_SOFTWARE = 0x8022;
    public static final int ATTR_RESPONSE_ORIGIN = 0x802B;
    public static final int ATTR_OTHER_ADDRESS = 0x802C;
    public static final int ATTR_FINGERPRINT = 0x8028;

    public static final int TRANSPORT_UDP = 17;
    private static final SecureRandom SECURE_RANDOM = new SecureRandom();

    private final int type;
    private final byte[] transactionId;
    private final List<Attribute> attributes;

    public StunMessage(int type, byte[] transactionId, List<Attribute> attributes) {
        this.type = type & 0xFFFF;
        this.transactionId = normalizeTransactionId(transactionId);
        this.attributes = attributes == null ? List.of() : List.copyOf(attributes);
    }

    public static StunMessage of(int type, byte[] transactionId, Attribute... attributes) {
        return new StunMessage(type, transactionId, attributes == null ? List.of() : List.of(attributes));
    }

    public static byte[] newTransactionId() {
        byte[] bytes = new byte[TRANSACTION_ID_BYTES];
        SECURE_RANDOM.nextBytes(bytes);
        return bytes;
    }

    public static boolean looksLike(byte[] packet, int offset, int length) {
        if (packet == null || offset < 0 || length < HEADER_BYTES || offset + HEADER_BYTES > packet.length) {
            return false;
        }
        if ((packet[offset] & 0xC0) != 0) {
            return false;
        }
        int declaredLength = Short.toUnsignedInt(ByteBuffer.wrap(packet, offset + 2, Short.BYTES).getShort());
        int cookie = ByteBuffer.wrap(packet, offset + 4, Integer.BYTES).getInt();
        return cookie == MAGIC_COOKIE && declaredLength + HEADER_BYTES <= length;
    }

    public static StunMessage parse(byte[] packet, int offset, int length) {
        if (!looksLike(packet, offset, length)) {
            return null;
        }
        ByteBuffer buffer = ByteBuffer.wrap(packet, offset, length);
        int type = Short.toUnsignedInt(buffer.getShort());
        int messageLength = Short.toUnsignedInt(buffer.getShort());
        int cookie = buffer.getInt();
        if (cookie != MAGIC_COOKIE || messageLength + HEADER_BYTES > length) {
            return null;
        }
        byte[] transactionId = new byte[TRANSACTION_ID_BYTES];
        buffer.get(transactionId);

        int end = offset + HEADER_BYTES + messageLength;
        List<Attribute> attributes = new ArrayList<>();
        while (buffer.position() < end) {
            if (end - buffer.position() < 4) {
                return null;
            }
            int attrType = Short.toUnsignedInt(buffer.getShort());
            int attrLength = Short.toUnsignedInt(buffer.getShort());
            if (attrLength > end - buffer.position()) {
                return null;
            }
            byte[] value = new byte[attrLength];
            buffer.get(value);
            int padding = padding(attrLength);
            if (padding > 0) {
                if (padding > end - buffer.position()) {
                    return null;
                }
                buffer.position(buffer.position() + padding);
            }
            attributes.add(new Attribute(attrType, value));
        }
        return new StunMessage(type, transactionId, attributes);
    }

    public byte[] toBytes() {
        return toBytes(null);
    }

    public byte[] toBytes(byte[] messageIntegrityKey) {
        int attributeBytes = 0;
        for (Attribute attribute : attributes) {
            attributeBytes += 4 + attribute.value().length + padding(attribute.value().length);
        }
        if (messageIntegrityKey != null && messageIntegrityKey.length > 0) {
            byte[] beforeIntegrity = serialize(attributeBytes + 24, attributes);
            byte[] digest = hmacSha1(messageIntegrityKey, beforeIntegrity);
            ByteBuffer packet = ByteBuffer.allocate(beforeIntegrity.length + 24);
            packet.put(beforeIntegrity);
            packet.putShort((short) ATTR_MESSAGE_INTEGRITY);
            packet.putShort((short) digest.length);
            packet.put(digest);
            return packet.array();
        }
        return serialize(attributeBytes, attributes);
    }

    private byte[] serialize(int declaredAttributeBytes, List<Attribute> serializedAttributes) {
        int actualAttributeBytes = 0;
        for (Attribute attribute : serializedAttributes) {
            actualAttributeBytes += 4 + attribute.value().length + padding(attribute.value().length);
        }
        ByteBuffer buffer = ByteBuffer.allocate(HEADER_BYTES + actualAttributeBytes);
        buffer.putShort((short) type);
        buffer.putShort((short) declaredAttributeBytes);
        buffer.putInt(MAGIC_COOKIE);
        buffer.put(transactionId);
        for (Attribute attribute : serializedAttributes) {
            buffer.putShort((short) attribute.type());
            buffer.putShort((short) attribute.value().length);
            buffer.put(attribute.value());
            for (int i = 0; i < padding(attribute.value().length); i++) {
                buffer.put((byte) 0);
            }
        }
        return buffer.array();
    }

    public static boolean verifyMessageIntegrity(byte[] packet, int offset, int length, byte[] messageIntegrityKey) {
        if (messageIntegrityKey == null || messageIntegrityKey.length == 0 || !looksLike(packet, offset, length)) {
            return false;
        }
        int declaredLength = Short.toUnsignedInt(ByteBuffer.wrap(packet, offset + 2, Short.BYTES).getShort());
        int end = offset + HEADER_BYTES + declaredLength;
        int position = offset + HEADER_BYTES;
        while (position < end) {
            if (end - position < 4) {
                return false;
            }
            int attrType = Short.toUnsignedInt(ByteBuffer.wrap(packet, position, Short.BYTES).getShort());
            int attrLength = Short.toUnsignedInt(ByteBuffer.wrap(packet, position + 2, Short.BYTES).getShort());
            int valueOffset = position + 4;
            int next = valueOffset + attrLength + padding(attrLength);
            if (attrLength > end - valueOffset || next > end) {
                return false;
            }
            if (attrType == ATTR_MESSAGE_INTEGRITY) {
                if (attrLength != 20) {
                    return false;
                }
                int signedLength = position - offset + 24 - HEADER_BYTES;
                byte[] signed = Arrays.copyOfRange(packet, offset, position);
                ByteBuffer.wrap(signed, 2, Short.BYTES).putShort((short) signedLength);
                byte[] expected = hmacSha1(messageIntegrityKey, signed);
                byte[] actual = Arrays.copyOfRange(packet, valueOffset, valueOffset + attrLength);
                return Arrays.equals(expected, actual);
            }
            position = next;
        }
        return false;
    }

    public int type() {
        return type;
    }

    public byte[] transactionId() {
        return Arrays.copyOf(transactionId, transactionId.length);
    }

    public String transactionIdHex() {
        return HexFormat.of().formatHex(transactionId);
    }

    public List<Attribute> attributes() {
        return attributes;
    }

    public Optional<Attribute> first(int type) {
        return attributes.stream().filter(attribute -> attribute.type() == type).findFirst();
    }

    public boolean hasAttribute(int type) {
        return attributes.stream().anyMatch(attribute -> attribute.type() == type);
    }

    public List<Attribute> all(int type) {
        return attributes.stream().filter(attribute -> attribute.type() == type).toList();
    }

    public Optional<InetSocketAddress> mappedAddress() {
        return first(ATTR_MAPPED_ADDRESS).flatMap(this::decodeAddress);
    }

    public Optional<InetSocketAddress> xorMappedAddress() {
        return first(ATTR_XOR_MAPPED_ADDRESS).flatMap(this::decodeXorAddress);
    }

    public Optional<InetSocketAddress> xorRelayedAddress() {
        return first(ATTR_XOR_RELAYED_ADDRESS).flatMap(this::decodeXorAddress);
    }

    public Optional<InetSocketAddress> xorPeerAddress() {
        return first(ATTR_XOR_PEER_ADDRESS).flatMap(this::decodeXorAddress);
    }

    public Optional<InetSocketAddress> responseOrigin() {
        return first(ATTR_RESPONSE_ORIGIN).flatMap(this::decodeAddress);
    }

    public Optional<InetSocketAddress> otherAddress() {
        return first(ATTR_OTHER_ADDRESS).flatMap(this::decodeAddress);
    }

    public Optional<InetSocketAddress> legacyXorResponseOrigin() {
        return first(ATTR_RESPONSE_ORIGIN).flatMap(this::decodeXorAddress);
    }

    public Optional<InetSocketAddress> legacyXorOtherAddress() {
        return first(ATTR_OTHER_ADDRESS).flatMap(this::decodeXorAddress);
    }

    public Optional<ChangeRequest> changeRequest() {
        return first(ATTR_CHANGE_REQUEST)
                .filter(attribute -> attribute.value().length == Integer.BYTES)
                .map(attribute -> {
                    int flags = ByteBuffer.wrap(attribute.value()).getInt();
                    return new ChangeRequest((flags & 0x04) != 0, (flags & 0x02) != 0);
                });
    }

    public int errorCode() {
        return first(ATTR_ERROR_CODE)
                .filter(attribute -> attribute.value().length >= 4)
                .map(attribute -> {
                    byte[] value = attribute.value();
                    return (value[2] & 0x07) * 100 + (value[3] & 0xFF);
                })
                .orElse(-1);
    }

    public List<Integer> unknownAttributes() {
        return first(ATTR_UNKNOWN_ATTRIBUTES)
                .map(attribute -> {
                    byte[] value = attribute.value();
                    List<Integer> result = new ArrayList<>(value.length / Short.BYTES);
                    ByteBuffer buffer = ByteBuffer.wrap(value);
                    while (buffer.remaining() >= Short.BYTES) {
                        result.add(Short.toUnsignedInt(buffer.getShort()));
                    }
                    return List.copyOf(result);
                })
                .orElseGet(List::of);
    }

    public Optional<byte[]> data() {
        return first(ATTR_DATA).map(attribute -> Arrays.copyOf(attribute.value(), attribute.value().length));
    }

    public Optional<String> username() {
        return textAttribute(ATTR_USERNAME);
    }

    public Optional<String> realm() {
        return textAttribute(ATTR_REALM);
    }

    public Optional<String> nonce() {
        return textAttribute(ATTR_NONCE);
    }

    public long lifetimeSeconds(long fallback) {
        return first(ATTR_LIFETIME)
                .filter(attribute -> attribute.value().length == Integer.BYTES)
                .map(attribute -> Integer.toUnsignedLong(ByteBuffer.wrap(attribute.value()).getInt()))
                .orElse(fallback);
    }

    public boolean requestedUdpTransport() {
        return first(ATTR_REQUESTED_TRANSPORT)
                .filter(attribute -> attribute.value().length >= 1)
                .map(attribute -> (attribute.value()[0] & 0xFF) == TRANSPORT_UDP)
                .orElse(false);
    }

    private Optional<InetSocketAddress> decodeXorAddress(Attribute attribute) {
        byte[] value = attribute.value();
        if (value.length != 8 && value.length != 20) {
            return Optional.empty();
        }
        int family = value[1] & 0xFF;
        int port = (((value[2] & 0xFF) << 8) | (value[3] & 0xFF)) ^ (MAGIC_COOKIE >>> 16);
        try {
            byte[] address;
            if (family == 0x01 && value.length >= 8) {
                int mask = MAGIC_COOKIE;
                address = new byte[4];
                for (int i = 0; i < 4; i++) {
                    address[i] = (byte) (value[4 + i] ^ ((mask >>> (24 - i * 8)) & 0xFF));
                }
            } else if (family == 0x02 && value.length >= 20) {
                byte[] mask = new byte[16];
                ByteBuffer.wrap(mask).putInt(MAGIC_COOKIE).put(transactionId);
                address = new byte[16];
                for (int i = 0; i < 16; i++) {
                    address[i] = (byte) (value[4 + i] ^ mask[i]);
                }
            } else {
                return Optional.empty();
            }
            return Optional.of(new InetSocketAddress(InetAddress.getByAddress(address), port));
        } catch (Exception e) {
            return Optional.empty();
        }
    }

    private Optional<InetSocketAddress> decodeAddress(Attribute attribute) {
        byte[] value = attribute.value();
        if (value.length != 8 && value.length != 20) {
            return Optional.empty();
        }
        int family = value[1] & 0xFF;
        int port = Short.toUnsignedInt(ByteBuffer.wrap(value, 2, Short.BYTES).getShort());
        try {
            byte[] address;
            if (family == 0x01 && value.length == 8) {
                address = Arrays.copyOfRange(value, 4, 8);
            } else if (family == 0x02 && value.length == 20) {
                address = Arrays.copyOfRange(value, 4, 20);
            } else {
                return Optional.empty();
            }
            return Optional.of(new InetSocketAddress(InetAddress.getByAddress(address), port));
        } catch (Exception e) {
            return Optional.empty();
        }
    }

    public static Attribute mappedAddress(InetSocketAddress address) {
        return new Attribute(ATTR_MAPPED_ADDRESS, encodeAddress(address));
    }

    public static Attribute xorMappedAddress(InetSocketAddress address, byte[] transactionId) {
        return new Attribute(ATTR_XOR_MAPPED_ADDRESS, encodeXorAddress(address, transactionId));
    }

    public static Attribute xorRelayedAddress(InetSocketAddress address, byte[] transactionId) {
        return new Attribute(ATTR_XOR_RELAYED_ADDRESS, encodeXorAddress(address, transactionId));
    }

    public static Attribute xorPeerAddress(InetSocketAddress address, byte[] transactionId) {
        return new Attribute(ATTR_XOR_PEER_ADDRESS, encodeXorAddress(address, transactionId));
    }

    public static Attribute responseOrigin(InetSocketAddress address) {
        return new Attribute(ATTR_RESPONSE_ORIGIN, encodeAddress(address));
    }

    public static Attribute otherAddress(InetSocketAddress address) {
        return new Attribute(ATTR_OTHER_ADDRESS, encodeAddress(address));
    }

    public static Attribute changeRequest(boolean changeIp, boolean changePort) {
        int flags = (changeIp ? 0x04 : 0) | (changePort ? 0x02 : 0);
        return new Attribute(ATTR_CHANGE_REQUEST, ByteBuffer.allocate(Integer.BYTES).putInt(flags).array());
    }

    public static Attribute unknownAttributes(int... types) {
        int[] normalized = types == null ? new int[0] : types;
        ByteBuffer buffer = ByteBuffer.allocate(normalized.length * Short.BYTES);
        for (int type : normalized) {
            buffer.putShort((short) (type & 0xFFFF));
        }
        return new Attribute(ATTR_UNKNOWN_ATTRIBUTES, buffer.array());
    }

    public static Attribute data(byte[] payload) {
        return new Attribute(ATTR_DATA, payload == null ? new byte[0] : Arrays.copyOf(payload, payload.length));
    }

    public static Attribute lifetime(long seconds) {
        ByteBuffer buffer = ByteBuffer.allocate(Integer.BYTES);
        buffer.putInt((int) Math.max(0, Math.min(Integer.toUnsignedLong(-1), seconds)));
        return new Attribute(ATTR_LIFETIME, buffer.array());
    }

    public static Attribute requestedUdpTransportAttribute() {
        return new Attribute(ATTR_REQUESTED_TRANSPORT, new byte[]{(byte) TRANSPORT_UDP, 0, 0, 0});
    }

    public static Attribute software(String value) {
        return new Attribute(ATTR_SOFTWARE, text(value));
    }

    public static Attribute username(String value) {
        return new Attribute(ATTR_USERNAME, text(value));
    }

    public static Attribute realm(String value) {
        return new Attribute(ATTR_REALM, text(value));
    }

    public static Attribute nonce(String value) {
        return new Attribute(ATTR_NONCE, text(value));
    }

    public static Attribute errorCode(int code, String reason) {
        int klass = Math.max(3, Math.min(6, code / 100));
        int number = Math.max(0, Math.min(99, code % 100));
        byte[] reasonBytes = text(reason == null ? "" : reason);
        ByteBuffer buffer = ByteBuffer.allocate(4 + reasonBytes.length);
        buffer.put((byte) 0);
        buffer.put((byte) 0);
        buffer.put((byte) klass);
        buffer.put((byte) number);
        buffer.put(reasonBytes);
        return new Attribute(ATTR_ERROR_CODE, buffer.array());
    }

    public static byte[] encodeAddress(InetSocketAddress address) {
        if (address == null || address.getAddress() == null) {
            throw new IllegalArgumentException("address is required");
        }
        byte[] raw = address.getAddress().getAddress();
        byte family = switch (raw.length) {
            case 4 -> 0x01;
            case 16 -> 0x02;
            default -> throw new IllegalArgumentException("unsupported address family");
        };
        ByteBuffer buffer = ByteBuffer.allocate(raw.length == 4 ? 8 : 20);
        buffer.put((byte) 0);
        buffer.put(family);
        buffer.putShort((short) address.getPort());
        buffer.put(raw);
        return buffer.array();
    }

    public static byte[] encodeXorAddress(InetSocketAddress address, byte[] transactionId) {
        if (address == null || address.getAddress() == null) {
            throw new IllegalArgumentException("address is required");
        }
        byte[] raw = address.getAddress().getAddress();
        byte family = switch (raw.length) {
            case 4 -> 0x01;
            case 16 -> 0x02;
            default -> throw new IllegalArgumentException("unsupported address family");
        };
        byte[] normalizedTransaction = normalizeTransactionId(transactionId);
        ByteBuffer buffer = ByteBuffer.allocate(raw.length == 4 ? 8 : 20);
        buffer.put((byte) 0);
        buffer.put(family);
        buffer.putShort((short) (address.getPort() ^ (MAGIC_COOKIE >>> 16)));
        if (raw.length == 4) {
            int mask = MAGIC_COOKIE;
            for (int i = 0; i < 4; i++) {
                buffer.put((byte) (raw[i] ^ ((mask >>> (24 - i * 8)) & 0xFF)));
            }
        } else {
            byte[] mask = new byte[16];
            ByteBuffer.wrap(mask).putInt(MAGIC_COOKIE).put(normalizedTransaction);
            for (int i = 0; i < 16; i++) {
                buffer.put((byte) (raw[i] ^ mask[i]));
            }
        }
        return buffer.array();
    }

    public static String hex(byte[] transactionId) {
        return HexFormat.of().formatHex(normalizeTransactionId(transactionId));
    }

    private static byte[] normalizeTransactionId(byte[] transactionId) {
        if (transactionId == null || transactionId.length != TRANSACTION_ID_BYTES) {
            byte[] generated = new byte[TRANSACTION_ID_BYTES];
            SECURE_RANDOM.nextBytes(generated);
            return generated;
        }
        return Arrays.copyOf(transactionId, transactionId.length);
    }

    private static int padding(int length) {
        return (4 - (length % 4)) % 4;
    }

    private static byte[] text(String value) {
        return (value == null ? "" : value).getBytes(StandardCharsets.UTF_8);
    }

    private Optional<String> textAttribute(int type) {
        return first(type).map(attribute -> new String(attribute.value(), StandardCharsets.UTF_8));
    }

    private static byte[] hmacSha1(byte[] key, byte[] payload) {
        try {
            Mac mac = Mac.getInstance("HmacSHA1");
            mac.init(new SecretKeySpec(key, "HmacSHA1"));
            return mac.doFinal(payload);
        } catch (Exception e) {
            throw new IllegalStateException("cannot compute STUN message integrity", e);
        }
    }

    public record Attribute(int type, byte[] value) {
        public Attribute {
            value = value == null ? new byte[0] : Arrays.copyOf(value, value.length);
        }
    }

    public record ChangeRequest(boolean changeIp, boolean changePort) {
        public static final ChangeRequest NONE = new ChangeRequest(false, false);
    }
}

File: implementations/csharp/server/src/Specus.StunServer/StunMessage.cs
using System.Buffers.Binary;
using System.Net;
using System.Net.Sockets;
using System.Security.Cryptography;
using System.Text;

namespace Specus.StunServer;

public sealed record StunAttribute(ushort Type, byte[] Value)
{
    public byte[] Value { get; init; } = Value.ToArray();
}

public readonly record struct ChangeRequest(bool ChangeIp, bool ChangePort);

public sealed class StunMessage
{
    public const uint MagicCookie = 0x2112A442;
    public const int HeaderBytes = 20;
    public const int TransactionIdBytes = 12;

    public const ushort BindingRequest = 0x0001;
    public const ushort BindingSuccess = 0x0101;
    public const ushort BindingError = 0x0111;

    public const ushort AttrMappedAddress = 0x0001;
    public const ushort AttrChangeRequest = 0x0003;
    public const ushort AttrErrorCode = 0x0009;
    public const ushort AttrUnknownAttributes = 0x000A;
    public const ushort AttrXorMappedAddress = 0x0020;
    public const ushort AttrPadding = 0x0026;
    public const ushort AttrResponsePort = 0x0027;
    public const ushort AttrSoftware = 0x8022;
    public const ushort AttrResponseOrigin = 0x802B;
    public const ushort AttrOtherAddress = 0x802C;

    public StunMessage(ushort type, byte[] transactionId, IReadOnlyList<StunAttribute>? attributes = null)
    {
        Type = type;
        TransactionId = transactionId.Length == TransactionIdBytes
            ? transactionId.ToArray()
            : throw new ArgumentException("transaction ID must contain 12 bytes", nameof(transactionId));
        Attributes = attributes?.Select(item => new StunAttribute(item.Type, item.Value)).ToList() ?? [];
    }

    public ushort Type { get; }
    public byte[] TransactionId { get; }
    public IReadOnlyList<StunAttribute> Attributes { get; }

    public static byte[] NewTransactionId()
    {
        var value = new byte[TransactionIdBytes];
        RandomNumberGenerator.Fill(value);
        return value;
    }

    public static StunMessage? Parse(ReadOnlySpan<byte> packet)
    {
        if (packet.Length < HeaderBytes
            || (packet[0] & 0xC0) != 0
            || BinaryPrimitives.ReadUInt32BigEndian(packet[4..8]) != MagicCookie)
        {
            return null;
        }
        var messageLength = BinaryPrimitives.ReadUInt16BigEndian(packet[2..4]);
        var end = HeaderBytes + messageLength;
        if (end > packet.Length)
        {
            return null;
        }
        var transactionId = packet[8..20].ToArray();
        var attributes = new List<StunAttribute>();
        for (var offset = HeaderBytes; offset < end;)
        {
            if (end - offset < 4)
            {
                return null;
            }
            var type = BinaryPrimitives.ReadUInt16BigEndian(packet[offset..(offset + 2)]);
            var length = BinaryPrimitives.ReadUInt16BigEndian(packet[(offset + 2)..(offset + 4)]);
            offset += 4;
            if (length > end - offset)
            {
                return null;
            }
            attributes.Add(new StunAttribute(type, packet.Slice(offset, length).ToArray()));
            offset += length + AlignmentPadding(length);
            if (offset > end)
            {
                return null;
            }
        }
        return new StunMessage(
            BinaryPrimitives.ReadUInt16BigEndian(packet[..2]),
            transactionId,
            attributes);
    }

    public byte[] ToBytes()
    {
        var attributeBytes = Attributes.Sum(
            attribute => 4 + attribute.Value.Length + AlignmentPadding(attribute.Value.Length));
        if (attributeBytes > ushort.MaxValue)
        {
            throw new InvalidOperationException("STUN message attributes exceed 65535 bytes");
        }
        var packet = new byte[HeaderBytes + attributeBytes];
        BinaryPrimitives.WriteUInt16BigEndian(packet.AsSpan(0, 2), Type);
        BinaryPrimitives.WriteUInt16BigEndian(packet.AsSpan(2, 2), (ushort)attributeBytes);
        BinaryPrimitives.WriteUInt32BigEndian(packet.AsSpan(4, 4), MagicCookie);
        TransactionId.CopyTo(packet.AsSpan(8, TransactionIdBytes));
        var offset = HeaderBytes;
        foreach (var attribute in Attributes)
        {
            if (attribute.Value.Length > ushort.MaxValue)
            {
                throw new InvalidOperationException(
                    $"STUN attribute 0x{attribute.Type:x4} exceeds 65535 bytes");
            }
            BinaryPrimitives.WriteUInt16BigEndian(packet.AsSpan(offset, 2), attribute.Type);
            BinaryPrimitives.WriteUInt16BigEndian(
                packet.AsSpan(offset + 2, 2),
                (ushort)attribute.Value.Length);
            offset += 4;
            attribute.Value.CopyTo(packet.AsSpan(offset, attribute.Value.Length));
            offset += attribute.Value.Length + AlignmentPadding(attribute.Value.Length);
        }
        return packet;
    }

    public StunAttribute? First(ushort type) => Attributes.FirstOrDefault(item => item.Type == type);

    public bool Has(ushort type) => Attributes.Any(item => item.Type == type);

    public ChangeRequest? ChangeRequestValue()
    {
        var value = First(AttrChangeRequest)?.Value;
        if (value is not { Length: 4 })
        {
            return null;
        }
        var flags = BinaryPrimitives.ReadUInt32BigEndian(value);
        return new ChangeRequest((flags & 0x04) != 0, (flags & 0x02) != 0);
    }

    public int? ResponsePortValue()
    {
        var value = First(AttrResponsePort)?.Value;
        return value is { Length: 2 }
            ? BinaryPrimitives.ReadUInt16BigEndian(value)
            : null;
    }

    public int ErrorCodeValue()
    {
        var value = First(AttrErrorCode)?.Value;
        return value is { Length: >= 4 }
            ? (value[2] & 0x07) * 100 + value[3]
            : -1;
    }

    public IPEndPoint? MappedAddressValue() => DecodeAddress(First(AttrMappedAddress)?.Value);
    public IPEndPoint? XorMappedAddressValue() => DecodeXorAddress(First(AttrXorMappedAddress)?.Value);
    public IPEndPoint? ResponseOriginValue() => DecodeAddress(First(AttrResponseOrigin)?.Value);
    public IPEndPoint? OtherAddressValue() => DecodeAddress(First(AttrOtherAddress)?.Value);

    public static StunAttribute MappedAddress(IPEndPoint endpoint) =>
        new(AttrMappedAddress, EncodeAddress(endpoint));

    public static StunAttribute XorMappedAddress(IPEndPoint endpoint, byte[] transactionId) =>
        new(AttrXorMappedAddress, EncodeXorAddress(endpoint, transactionId));

    public static StunAttribute ResponseOrigin(IPEndPoint endpoint) =>
        new(AttrResponseOrigin, EncodeAddress(endpoint));

    public static StunAttribute OtherAddress(IPEndPoint endpoint) =>
        new(AttrOtherAddress, EncodeAddress(endpoint));

    public static StunAttribute ChangeRequestAttribute(bool changeIp, bool changePort)
    {
        var value = new byte[4];
        BinaryPrimitives.WriteUInt32BigEndian(
            value,
            (changeIp ? 0x04U : 0U) | (changePort ? 0x02U : 0U));
        return new StunAttribute(AttrChangeRequest, value);
    }

    public static StunAttribute ResponsePort(int port)
    {
        if (port is < 0 or > 65535)
        {
            throw new ArgumentOutOfRangeException(nameof(port));
        }
        var value = new byte[2];
        BinaryPrimitives.WriteUInt16BigEndian(value, (ushort)port);
        return new StunAttribute(AttrResponsePort, value);
    }

    public static StunAttribute Padding(int length) =>
        new(AttrPadding, new byte[Math.Clamp(length, 0, 65_503)]);

    public static StunAttribute Software(string value) =>
        new(AttrSoftware, Encoding.UTF8.GetBytes(value));

    public static StunAttribute ErrorCode(int code, string reason)
    {
        var klass = Math.Clamp(code / 100, 3, 6);
        var number = Math.Clamp(code % 100, 0, 99);
        var reasonBytes = Encoding.UTF8.GetBytes(reason);
        var value = new byte[4 + reasonBytes.Length];
        value[2] = (byte)klass;
        value[3] = (byte)number;
        reasonBytes.CopyTo(value.AsSpan(4));
        return new StunAttribute(AttrErrorCode, value);
    }

    public static StunAttribute UnknownAttributes(params ushort[] types)
    {
        var value = new byte[types.Length * 2];
        for (var index = 0; index < types.Length; index++)
        {
            BinaryPrimitives.WriteUInt16BigEndian(value.AsSpan(index * 2, 2), types[index]);
        }
        return new StunAttribute(AttrUnknownAttributes, value);
    }

    internal static byte[] XorAddressValue(IPEndPoint endpoint, byte[] transactionId) =>
        EncodeXorAddress(endpoint, transactionId);

    private static byte[] EncodeAddress(IPEndPoint endpoint)
    {
        var address = endpoint.Address.AddressFamily == AddressFamily.InterNetworkV6
            ? endpoint.Address.GetAddressBytes()
            : endpoint.Address.MapToIPv4().GetAddressBytes();
        var value = new byte[address.Length == 4 ? 8 : 20];
        value[1] = address.Length == 4 ? (byte)0x01 : (byte)0x02;
        BinaryPrimitives.WriteUInt16BigEndian(value.AsSpan(2, 2), (ushort)endpoint.Port);
        address.CopyTo(value.AsSpan(4));
        return value;
    }

    private static IPEndPoint? DecodeAddress(byte[]? value)
    {
        if (value is null || (value.Length != 8 && value.Length != 20))
        {
            return null;
        }
        var port = BinaryPrimitives.ReadUInt16BigEndian(value.AsSpan(2, 2));
        return value[1] switch
        {
            0x01 when value.Length == 8 => new IPEndPoint(new IPAddress(value.AsSpan(4, 4)), port),
            0x02 when value.Length == 20 => new IPEndPoint(new IPAddress(value.AsSpan(4, 16)), port),
            _ => null,
        };
    }

    private static byte[] EncodeXorAddress(IPEndPoint endpoint, byte[] transactionId)
    {
        var value = EncodeAddress(endpoint);
        var port = BinaryPrimitives.ReadUInt16BigEndian(value.AsSpan(2, 2));
        BinaryPrimitives.WriteUInt16BigEndian(
            value.AsSpan(2, 2),
            (ushort)(port ^ (MagicCookie >> 16)));
        Span<byte> mask = stackalloc byte[16];
        BinaryPrimitives.WriteUInt32BigEndian(mask[..4], MagicCookie);
        transactionId.CopyTo(mask[4..]);
        for (var index = 4; index < value.Length; index++)
        {
            value[index] ^= mask[index - 4];
        }
        return value;
    }

    private IPEndPoint? DecodeXorAddress(byte[]? value)
    {
        if (value is null || (value.Length != 8 && value.Length != 20))
        {
            return null;
        }
        var decoded = value.ToArray();
        var port = BinaryPrimitives.ReadUInt16BigEndian(decoded.AsSpan(2, 2));
        BinaryPrimitives.WriteUInt16BigEndian(
            decoded.AsSpan(2, 2),
            (ushort)(port ^ (MagicCookie >> 16)));
        Span<byte> mask = stackalloc byte[16];
        BinaryPrimitives.WriteUInt32BigEndian(mask[..4], MagicCookie);
        TransactionId.CopyTo(mask[4..]);
        for (var index = 4; index < decoded.Length; index++)
        {
            decoded[index] ^= mask[index - 4];
        }
        return DecodeAddress(decoded);
    }

    private static int AlignmentPadding(int length) => (4 - length % 4) % 4;
}
